### PR TITLE
[inductor] Stop using `x + tl.zeros(...)` in generated triton

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99661
* #100569
* __->__ #100163
* #100573
* #100572

For reductions, this changes the accumulator
```python
_tmp2 = tl.zeros([XBLOCK, RBLOCK], tl.int8) + -128
```
to
```python
_tmp2 = tl.full([XBLOCK, RBLOCK], -128, tl.int32)
```
which is equivalent since addition does type promotion from `int8` to `int32`

For constant indexing, this changes
```python
tl.store(in_out_ptr0 + (0 + tl.zeros([XBLOCK, 1], tl.int32)), tmp4, None)
```
to
```python
tl.store(in_out_ptr0 + (tl.full([XBLOCK, 1], 0, tl.int32)), tmp4, None)
```

For variable indexing, this changes
```python
tl.store(out_ptr0 + (0 + tl.zeros([XBLOCK], tl.int32)), tmp1, None)
```
to
```python
tl.store(out_ptr0 + (tl.broadcast_to(x0, [XBLOCK])), tmp1, None)
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -1259,7 +1260,14 @@ def final_argreduce(buffer, result_var, value, index):
                 self.compute.writeline(
                     f"{accumulator} = tl.where({cond}, {updated}, {accumulator})"
                 )
-                self.suffix.writeline(f"{result_var} = {final_reduction(accumulator)}")
+
+                if src_dtype == torch.bool:
+                    # tl.reduce doesn't support tl.int1
+                    accumulator = f"{accumulator}.to(tl.int8)"
+                    result_type = triton_compute_type(dtype)
+                    self.suffix.writeline(f"{result_var} = {final_reduction(accumulator)}.to({result_type})")
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:1268]: This is only really used for `aten.any`. It changes the final reduction of a non-persistent reduction from
```python
tmp5 = triton_helpers.max(_tmp5, 1)[:, None]
```

```python
tmp5 = triton_helpers.max(_tmp5.to(tl.int8), 1)[:, None].to(tl.int1)
```

Comment (ngimel) [reply torch/_inductor/codegen/triton.py:1268]: This is good to put as a comment in the code!

**Pull request closed without merging**